src/schedule/ll_schedule.c

```diff
@@ -113,13 +119,39 @@ static void schedule_ll_task_done(struct ll_schedule_data *sch,
 		atomic_read(&sch->domain->total_num_tasks));
 }
 
+/* perf measurement windows size 2^x */
+#define CHECKS_WINDOW_SIZE	10
+
+static inline void dsp_load_check(struct ll_dsp_load *load, uint32_t cycles0, uint32_t cycles1)
+{
+	uint32_t diff, max;
+
+	if (cycles1 > cycles0)
+		diff = cycles1 - cycles0;
+	else
+		diff = UINT32_MAX - cycles0 + cycles1;
+
+	load->cycles_sum += diff;
+
+	if (++load->checks == 1 << CHECKS_WINDOW_SIZE) {
+		load->cycles_sum >>= CHECKS_WINDOW_SIZE;
+		max = clock_us_to_ticks(PLATFORM_DEFAULT_CLOCK, CONFIG_SYSTICK_PERIOD);
```

Comment:
This should be light-weight, but the above line performs run-time divisions and multiplications with constants, even if only every N runs. Since `clock_us_to_ticks()` is a real function, I don't think this will be pre-calculated and optimised away by the compiler. But since it's always the same calculation (as long as the clock frequency doesn't change), this should be calculated once and stored. To also take frequency calculation into account we can also store the frequency (index) that was used to calculate this, and if it changes - recalculate.
But yes, this would also somehow have to account for imprecise XTOS ticks.

Reply:
The only part doing the div/mult is the ```clock_us_to_ticks()``` and this will only happen once per second (and you are right, we could do this at init since it's all const), all the rest is shifts which are optimal on xtensa